teachertool/src/utils/index.ts

```diff
@@ -44,3 +45,24 @@ export function getProjectLink(inputText: string): string {
     const hasMakeCode = inputText?.indexOf("makecode") !== -1;
     return hasMakeCode ? inputText : `https://makecode.com/${inputText}`;
 }
+
+export function splitCriteriaTemplate(template: string): CriteriaTemplateSegment[] {
+    // Split by the regex, which will give us an array where every other element is a parameter.
+    const paramRegex = /\$\{([\w\s]+)\}/g;
+    const parts = template.split(paramRegex);
+
+    const segments: CriteriaTemplateSegment[] = [];
+    for (let i = 0; i < parts.length; i++) {
+        const part = parts[i];
+
+        if (part) {
+            segments.push({ type: i % 2 === 0 ? "plain-text" : "param", content: part });
```

Comment:
This loop is making several assumptions I'm not sure are safe to make:
* The `parts` array begins with a `plain-text` element.
* Two consecutive `param` elements will be separated by a `plain-text` element.

Are you certain these will hold true for all possible input?

Reply:
I did wonder about these when I was working on it, but the split with a regex separator inserts empty strings even between consecutive matches, [and at the start if the first element is a separator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/split#:~:text=If%20separator%20appears%20at%20the%20beginning%20(or%20end)%20of%20the%20string%2C%20it%20still%20has%20the%20effect%20of%20splitting%2C%20resulting%20in%20an%20empty%20(i.e.%20zero%20length)%20string%20appearing%20at%20the%20first%20(or%20last)%20position%20of%20the%20returned%20array.), so I am pretty confident this is a safe assumption.

For good measure, I tested out with "${one} ${two}${three}" and here's the result:
![image](https://github.com/microsoft/pxt/assets/69657545/d7444715-4ac6-4438-a778-33690bb2d4fd)
![image](https://github.com/microsoft/pxt/assets/69657545/9e60ebcf-17d4-4b18-8f67-b57940730972)
![image](https://github.com/microsoft/pxt/assets/69657545/8063bafb-d58f-4bb6-bd01-e0636dc9b579)

Reply:
I am surprised it works that well, but ok!

Reply:
You may want to trim the parsed param names. Consider the input `"${ param1 }   ${ param2 }". I think you'll end up with param names in in the `segments` array that include the padded whitespace.

Reply:
> I am surprised it works that well, but ok!

Happy to try out other test cases if you can think of any. I think I've verified all the ones that came to my mind, at least (begin with param, end with param, params next to each other, no params, one parameter and nothing else, and of course, params sprinkled throughout with plain text)

Reply:
Two that I can think of:
* Nested params: `${this is ${nested}} ${this_is_a_valid_param}`.
* Dot notation: `${user.name} ${user.email} ${this_is_a_valid_param}`.

Fine if nested/dotted params don't parse out as parameters, but would be good to know the parser can recover and successfully pull out a valid param appearing later.

Reply:
Hmm interesting. They definitely don't parse out as parameters, but it also doesn't crash and it is able to get the valid parameter afterwards.

Testing `${this is ${nested}} ${count}` and `some text ${user.name} ${user.email} ${count} some more text`, we get:
![image](https://github.com/microsoft/pxt/assets/69657545/8680f2b3-c0d3-4708-ba67-741eb7faeabc)
![image](https://github.com/microsoft/pxt/assets/69657545/0c847f88-5a44-49df-9222-79f58cb57aac)

Unless there are concerns from your end, I'm inclined to leave it be for now and worry about supporting those scenarios if/when we need them.

Reply:
Ya that's fine. Looks like it was able to find the last param. There are libraries that do this kind of thing really well, though probably overkill for us at this time. The [classic "handlbars" package](https://handlebarsjs.com/), comes to mind.